lib/msf/base/sessions/pingback.rb

```diff
@@ -0,0 +1,102 @@
+# -*- coding: binary -*-
+require 'msf/base'
+
+module Msf
+module Sessions
+
+###
+#
+# This class provides the ability to receive a pingback UUID
+#
+###
+class Pingback
+
+  #
+  # This interface supports basic interaction.
+  #
+  include Msf::Session
+  include Msf::Session::Basic
+
+  #
+  # Returns the type of session.
+  #
+  def self.type
+    "pingback"
+  end
+
+  def initialize(rstream, opts={})
+    super
+    self.platform ||= ""
+    self.arch     ||= ""
+    datastore = opts[:datastore]
+  end
+
+  def self.create_session(rstream, opts={})
+    Msf::Sessions::Pingback.new(rstream, opts)
+  end
+
+  def process_autoruns(datastore)
+    uuid_read
+    cleanup
+  end
+
+  def cleanup
+    if rstream
```

Comment:
Given the rescue and assignment that happen in here, this guard doesn't do any thing.

Reply:
It prevents unnecessary exception handling, when we already know the `close` method call is going to fail.

* https://simonecarletti.com/blog/2010/01/how-slow-are-ruby-exceptions/
* https://www.honeybadger.io/blog/benchmarking-exceptions-in-ruby-yep-theyre-slow/
* https://gist.github.com/JoshCheek/95ef47502f1976eb4a3c

Reply:
I definitely wouldn't want to generate unnecessary exceptions in a tight loop or hot code path. This isn't in one of those, though, and in my testing this method doesn't get called with a falsey `rstream` value (not that it never could be called that way or should require a non-null `rstream` as a precondition). This guard is optimizing for the exceptional case and thus is overly defensive. It doesn't _need_ to be one way or the other, I was just pointing out that it was extra code that could be safely removed.

Reply:
I'll note that this method is just a clone of the same from `lib/msf/base/sessions/command_shell.rb`. While I can appreciate further minimization, this may not be the only thing we ever use pingback payloads for (there may be other work we do on cleanup when there are other protocols implemented).